kv: add fixtures and verify downstream-of-raft proto types

tamird opened the pull request:
Somewhat a WIP, but definitely ready for a look.

The main downside is that this makes the `kv` tests take 56s, up from ... 44s? Whoa, when did our `kv` tests get this slow?

cc @cockroachdb/core

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6038)

<!-- Reviewable:end -->

tamird committed:
storage/engine: remove old optimization

All our protos have generated marshalling code, so this is dead code.

tamird committed:
Move proto utils into their own package

tamird committed:
storage: add fixtures and verify downstream-of-raft proto types

petermattis commented:
Review status: 0 of 40 files reviewed at latest revision, 8 unresolved discussions.

---

*[kv/main_test.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFfPdgPB8_TxlJ17it:-KFFfPdh7C1etwwNupMh:-1192629388) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L50)):*
Perhaps this should be moved into `TestMain` so that you don't need a `randGen` global. Or a separate `checkGoldenProtos` function.

---

*[kv/main_test.go, line 90 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFF_2Xef1ZJTUqDlNYI:-KFF_2Xef1ZJTUqDlNYJ:-1880304931) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L90)):*
Let's use sha1 or something like that.

---

*[storage/track_raft_protos.go, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFfhsEk__7RE7tR5Sr:-KFFfhsEk__7RE7tR5Ss:1502639918) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L1)):*
`s/2014/2016/g`

---

*[storage/track_raft_protos.go, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFk2uYsk7J9Q-AIceD:-KFFk2uZ2BQWdZU5UPE9:287284608) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L39)):*
Why are you also tracking gossip protos here? Add a comment because it isn't obvious why this is necessary.

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFFn7vCROfc0jdm1PI0:-57707515) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
The `bufferPool` was an important optimization at some point. Rather than removing it, can you add a `protoutil.TrackMarshal` call that simply does the tracking and leaves the marshaling as-is below.

---

*[util/protoutil/marshal.go, line 35 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFmFlQuHPMKSMGpflI:-KFFmFlQuHPMKSMGpflJ:1456586717) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/util/protoutil/marshal.go#L35)):*
Locking on every call to `protoutil.Marshal` is unfortunate. If the state were a single function pointer I think this could be reduced to an `atomic.Value`. Then we would only suffer the locking in tests and not in prod code.

---

*[util/protoutil/marshal.go, line 37 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFmUzDpK8rmUuIBCFM:-KFFmUzDpK8rmUuIBCFN:164951332) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/util/protoutil/marshal.go#L37)):*
Can't the `isTrack` function keep this map if it needs it?

---

*[util/protoutil/marshal.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFmZ8t9161lP3kI75M:-KFFmZ8t9161lP3kI75N:636395933) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/util/protoutil/marshal.go#L49)):*
You should panic if `tracker.isTrack != nil`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

tamird commented:
Moved this instrumentation into the storage package - turns out there were some cases not caught by the kv tests.

Also added "negative" verification - that is, the implementation now checks that all goldenProtos are indeed tickled by the tests....this might actually be annoying if folks are running just a few tests, so i'm open to suggestions here.

---

Review status: 0 of 40 files reviewed at latest revision, 8 unresolved discussions.

---

*[kv/main_test.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFfPdgPB8_TxlJ17it:-KFFsWqvQpPXuFg-XMVq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L50)):*
Done.

---

*[kv/main_test.go, line 90 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFF_2Xef1ZJTUqDlNYI:-KFFsfAa1bg8Z_-W0sYl:653539327) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L90)):*
Any particular justification?

---

*[storage/track_raft_protos.go, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFfhsEk__7RE7tR5Sr:-KFFstkFGKODNVY5K_BO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L1)):*
Done.

---

*[storage/track_raft_protos.go, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFk2uYsk7J9Q-AIceD:-KFFoZAk6_ikTkWWFSF-:-1498567730) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L39)):*
I am actually not tracking the gossip protos here - this function is grabbed so that it can be ignored (application of raft commands can trigger gossiping of system configs).

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFFolkDwdA8D80jCdX6:1659079407) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
I'm not sure what you mean - what tracking?

This change is in a separate commit which explains why it was done - this optimization is dead code.

---

*[util/protoutil/marshal.go, line 35 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFmFlQuHPMKSMGpflI:-KFFro6JTXRlCZaqTdr3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/util/protoutil/marshal.go#L35)):*
Done.

---

*[util/protoutil/marshal.go, line 37 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFmUzDpK8rmUuIBCFM:-KFFrjhDpBS3lyPSVRuA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/util/protoutil/marshal.go#L37)):*
Done.

---

*[util/protoutil/marshal.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFmZ8t9161lP3kI75M:-KFFrkEKeZJyHQCxG2I9:-1968779237) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/util/protoutil/marshal.go#L49)):*
This code has been obsoleted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-:-KFFsv6R2favL3CRQoJQ:-1090045357)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 41 of 42 files at r1, 1 of 1 files at r2.
Review status: 0 of 43 files reviewed at latest revision, 14 unresolved discussions.

---

*[kv/main_test.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFwMyVMOskPMMsVCnU:-KFFwMyW_5WL8mia6goA:632041658) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L52)):*
`NewPopulatedX` doesn't fill in every field: for some fields it has a 10% chance of leaving it blank. That might cause it to miss things. 

It also doesn't look like it would reliably catch the difference between a nullable and non-nullable field, which is one of our specific concerns. Maybe we should also  checksum a zero-valued protobuf? Encoding and checksumming (into a single checksum value) a zero-valued protobuf and say 20 `NewPopulated` ones would address both concerns.

---

*[kv/main_test.go, line 90 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFF_2Xef1ZJTUqDlNYI:-KFFzXMDM9sL7_6cjWTY:266348076) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L90)):*
Use at least `sha256` if you're going to use a cryptographic hash, or `fnv.New64a()` for a non-cryptographic one. I'm fine with a non-cryptographic hash in this context, but `fnv.New64a` appears to be recommended over the plain `fnv.New64`. 

---

*[kv/main_test.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG-uTGnlEeWJZXQDuM:-KFG-uTGnlEeWJZXQDuN:1768623381) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L106)):*
Failures at this point are not going to be reported well in circleci. I would most of the golden proto stuff into a regular test, and the test for "missing fixtures" in the hook function itself (the hook can panic if it finds we're downstream of raft and trying to marshal a proto that's not on the whitelist. This will also give better error reporting; otherwise it could be hard to figure out how exactly a new proto type is being encoded downstream of raft). Then the only thing that needs to happen in `TestMain` is installing (and optionally removing) the hook.

---

*[kv/main_test.go, line 17 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFu85TylJw6Qlmd18x:-KFFu85TylJw6Qlmd18y:958007240) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/kv/main_test.go#L17)):*
Package `kv` doesn't have complete test coverage (e.g. `TruncateLog` is only tested in package `storage`, which is why you haven't seen `TruncatedState` reported). We need this in at least package `storage`, and it would be best if we could include this (or maybe a more limited version) in every test package to make sure we're not missing anything that is only tested through a non-obvious path.

---

*[storage/track_raft_protos.go, line 18 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG0jzbOX6qtNLwTERh:-KFG0jzbOX6qtNLwTERi:-595080257) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L18)):*
Remove these names and add your own. 

---

*[storage/track_raft_protos.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG1c4vz42eX-AJJQHZ:-KFG1c4vz42eX-AJJQH_:-555287492) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L48)):*
Might be worth adding a new function to `util/caller` to return multiple caller's names (with caching). 

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFFsqamgEBIeCDwmWHZ:-1339274852) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
We could add a type assertion in `protoutil.Marshal` that the message implements `proto.Marshaler` (in which case the buffer optimization has no effect). If we make `TrackMarshal` a separate function it would be very hard to ensure that there are no untracked calls to `Marshal`.

---

*[util/protoutil/marshal.go, line 35 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFmFlQuHPMKSMGpflI:-KFFriOmGAZh5navSviH:-1906419762) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/util/protoutil/marshal.go#L35)):*
+1. This should just be a `marshalHook` function pointer and the function itself should do all the tracking.

---

*[util/uuid/uuid.go, line 98 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFr0MsypmHA4iSt0Wv:-KFFr0MsypmHA4iSt0Ww:924700576) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/util/uuid/uuid.go#L98)):*
`NewV4` generates random data which you just clobber with the pseudo-random data from the interface. It's probably better to just make a zero-valued UUID instead of calling `NewV4`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 43 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/main_test.go, line 52 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG7fezQvMyV4BbErjR:-KFG7fezQvMyV4BbErjS:-1095422942) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/main_test.go#L52)):*
This might be a level of unnecessary paranoia: how do we know that this testing is actually catching anything? I think there should be a test which makes a new proto message that is identical to one of the existing ones (such as `roachpb.Timestamp`) but which has the addition of a non-nullable field.

---

*[storage/track_raft_protos.go, line 64 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG5oMaQ9rjExTqaNx7:-KFG5oMaQ9rjExTqaNx8:-557783340) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/track_raft_protos.go#L64)):*
Add a comment explaining that this is ignoring protos marshaled by gossip which is sometimes invoked "below" raft. I misunderstood what was going on the first time I read this code. I'm sure someone else will do the same.

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFG61b0Ju-Ylg9i5YTL:-1885836227) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
Ah. I seem to have trouble looking at the commit messages when using reviewable. This has happened so often I'm going to blame them and not myself.

---

*[util/protoutil/marshal.go, line 26 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG6H2vJL9xQyos8xqA:-KFG6H2vJL9xQyos8xqB:734445865) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/util/protoutil/marshal.go#L26)):*
Sort of ugly that the caller has to know the function signature to store here. I'm surprised you didn't use a `SetInterceptor` approach instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 44 files reviewed at latest revision, 12 unresolved discussions.

---

*[kv/main_test.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFwMyVMOskPMMsVCnU:-KFG2uCUcoybLcW9wEiC:-245305645) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L52)):*
This is already checksumming a zero-valued protobuf; I've added the 20 `NewPopulated` calls as you suggested.

---

*[kv/main_test.go, line 90 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFF_2Xef1ZJTUqDlNYI:-KFGCx8uewqTfTogEhWM:-1727538180) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L90)):*
Yep, discovered this as well; changed to `New64a`.

---

*[kv/main_test.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG-uTGnlEeWJZXQDuM:-KFGF3h8xyHrMDoGhOLX:1324794133) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/kv/main_test.go#L106)):*
Good point; Done.

---

*[kv/main_test.go, line 17 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFu85TylJw6Qlmd18x:-KFG2eLB_hleFfNM7L34:613244450) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/kv/main_test.go#L17)):*
Yep, I moved it to storage, which seems to give a superset of the coverage.

Moving it to every package is possible, but perhaps cumbersome (it requires something akin to leaktest's generation script). Do you feel that it's needed?

---

*[storage/main_test.go, line 52 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG7fezQvMyV4BbErjR:-KFGB_sSYrO9_PwXgvU3:-230924469) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/main_test.go#L52)):*
I'm not following - what change to this test are you envisioning that would silently break it as you describe?

---

*[storage/track_raft_protos.go, line 18 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG0jzbOX6qtNLwTERh:-KFGBtjggw_1r4g13oKk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L18)):*
Done.

---

*[storage/track_raft_protos.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG1c4vz42eX-AJJQHZ:-KFGBmtYs_l0lAIui5cO:-930465704) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L48)):*
I'm going to leave that to the next person that needs it, if that's OK.

---

*[storage/track_raft_protos.go, line 64 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG5oMaQ9rjExTqaNx7:-KFGrDIZht5abZMEssZ0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/track_raft_protos.go#L64)):*
Done.

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFGCQTS-UWcAtO_hVFP:-1894845213) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
@bdarnell we could add that type assertion if you feel strongly. I'm not sure it's worth it since our generator takes care of that, and the assertion does have a minor associated cost.

---

*[util/protoutil/marshal.go, line 26 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG6H2vJL9xQyos8xqA:-KFGD5UBQiEb0xOcUj6u:-1622819128) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/util/protoutil/marshal.go#L26)):*
Good point. Done.

---

*[util/uuid/uuid.go, line 98 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFr0MsypmHA4iSt0Wv:-KFGCeAtZINLvjwxcRXM:-91416392) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/util/uuid/uuid.go#L98)):*
Sure. The intention was to not repeat the other initialization that is done in `NewV4`, but it doesn't seem necessary anyway.

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 44 files reviewed at latest revision, 11 unresolved discussions.

---

*[storage/main_test.go, line 52 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG7fezQvMyV4BbErjR:-KFHTGihYsgqLB3yfp0x:-1879082327) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/main_test.go#L52)):*
I'm envisioning a new test which uses a new proto that is for test purposes only:

```
message Timestamp2 {
  optional int64 wall_time = 1 [(gogoproto.nullable) = false];
  optional int32 logical = 2 [(gogoproto.nullable) = false];
  optional int32 dummy = 3 [(gogoproto.nullable) = false];
}
```

The test would then serialize `&Timestamp2{}` and a populated `&Timestamp2{}` and verify the hash differs from a `Timestamp` containing the same contents (modulo the `dummy` field). The purpose is to add paranoia that the verification against golden values is actually working correctly in an instance where we know it should find a problem. 

---

*[storage/track_raft_protos.go, line 64 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG5oMaQ9rjExTqaNx7:-KFHUDiMJWPM3Hn-2F89:1008910472) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/track_raft_protos.go#L64)):*
Did the comment get lost somewhere? I'm not seeing it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 44 files reviewed at latest revision, 11 unresolved discussions.

---

*[storage/main_test.go, line 52 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG7fezQvMyV4BbErjR:-KFI-MOZZEYz2RKALXB8:-1898005165) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/main_test.go#L52)):*
Sure, this is doable, but I'm not sure what it's demonstrating. We couldn't use the same code to exercise that case, so it's not increasing our confidence in _this_ code here; what's the upshot?

---

*[storage/track_raft_protos.go, line 64 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG5oMaQ9rjExTqaNx7:-KFI-CbPUYzS0Oin5eog:31834510) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/track_raft_protos.go#L64)):*
The comment is higher up, on the definition of `addInfoFunc`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 42 files at r1, 8 of 11 files at r4, 33 of 37 files at r5, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[roachpb/data.proto, line 27 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJ6SvlaXtCHlwNVvAD:-KFJ6SvmOLeZLdC7gcHx:1160258690) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/roachpb/data.proto#L27)):*
is the only difference to `populate_all` how it treats nested messages?

---

*[roachpb/data.proto, line 42 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJ6aAE9ZFDjNkNYE5r:-KFJ6aAE9ZFDjNkNYE5s:1862886893) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/roachpb/data.proto#L42)):*
Not worth just making it a default?

---

*[storage/below_raft_protos_test.go, line 45 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJ7kMASlrl1vIK1bwb:-KFJ7kMASlrl1vIK1bwc:39013740) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/storage/below_raft_protos_test.go#L45)):*
nit: if this (and `marshalerTo`) are only used in the methods right below, you can move the type def into the respective methods.

---

*[storage/below_raft_protos_test.go, line 61 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJEVPRqPXqgcy6-cqE:-KFJEVPRqPXqgcy6-cqF:-824550742) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/storage/below_raft_protos_test.go#L61)):*
Could you sprinkle some comments? This is a bit hard to parse, mostly because the expected sum is part of the data here, but the seed used isn't. Maybe it ought to be next to this as `const goldenSeed = 1337` (same for `goldenIterations`).

---

*[storage/main_test.go, line 52 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG7fezQvMyV4BbErjR:-KFJLFXXI_S-96QZakaM:587416049) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/main_test.go#L52)):*
Agreed that there could be a silly sanity test somewhere (probably close to the `goldenProtos` definition), but it wouldn't be here (after all, this is testing something else, namely that we're checking the right protos).

---

*[storage/main_test.go, line 39 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJGbObFMJxLUdl4Rzh:-KFJGbObFMJxLUdl4Rzi:1453400788) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/storage/main_test.go#L39)):*
Needs commentary from here on down.

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFJJUx13ONm5R8wpWiQ:319781262) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
Was it "dead code" in the sense that the generated marshaler already has that type of optimization? Does seem that the code was run, so I wouldn't call it "dead".

---

*[util/protoutil/marshal.go, line 25 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJK7PSqdsTgq4UW_E2:-KFJK7PTHp2ujiszf1xb:12228552) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/util/protoutil/marshal.go#L25)):*
 you even need an atomic here? This is testing-only (and used only from `TestMain` which together with leaktest can obviate the need for any synchronization here). If this here is free anyway, then I'm not opposed, but it sure is squarely in the critical path.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 40 of 44 files reviewed at latest revision, 17 unresolved discussions.

---

*[roachpb/data.proto, line 27 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJ6SvlaXtCHlwNVvAD:-KFK68AeH1GeWkxpZEgT:96320602) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/roachpb/data.proto#L27)):*
`populate_all` is a file-level option.

---

*[roachpb/data.proto, line 42 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJ6aAE9ZFDjNkNYE5r:-KFK6Ag_NzTQPeVRsHIL:-1552704735) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/roachpb/data.proto#L42)):*
Not possible - `populate` requires that all constituent types are also `populate`, and in the case of our protos which wrap raft protos, that condition is not met.

---

*[storage/below_raft_protos_test.go, line 45 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJ7kMASlrl1vIK1bwb:-KFK74UMDiFUqoL6kt42:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/storage/below_raft_protos_test.go#L45)):*
Done.

---

*[storage/below_raft_protos_test.go, line 61 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJEVPRqPXqgcy6-cqE:-KFK78651yWhISKrabdO:-1756720605) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/storage/below_raft_protos_test.go#L61)):*
Sounds good, moved.

---

*[storage/main_test.go, line 52 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG7fezQvMyV4BbErjR:-KFK7J_oH49v2ApNv1tV:-1434191505) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/main_test.go#L52)):*
@petermattis' comment was originally anchored on the "golden" test, which used to be in this file.

In any case, I'm still not convinced that there's any benefit to be had here.

---

*[storage/main_test.go, line 39 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJGbObFMJxLUdl4Rzh:-KFK7YfoSsM2tKf0lHIq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/storage/main_test.go#L39)):*
Done.

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFK4U1_qtEdM-x8KEw-:-967011572) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
It was dead code in that `proto.Buffer` is not used by any proto that has gogoproto's marshaler option. In fact, event `proto.(*Buffer).Marshal` checks to see if the proto implements `proto.Marshaler`, and if it does, simply delegates.

---

*[util/protoutil/marshal.go, line 25 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJK7PSqdsTgq4UW_E2:-KFK4uMU4ddt9iyofazN:-1622819128) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/util/protoutil/marshal.go#L25)):*
Good point. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. This is going to be the best kind of annoying: the useful kind.

---

Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[roachpb/data.proto, line 42 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJ6aAE9ZFDjNkNYE5r:-KFLU-qOT7H30wbcOWWA:-604095275) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/roachpb/data.proto#L42)):*
But you **could** just have populate methods generated for everything and chose not to, right?

---

*[storage/below_raft_protos_test.go, line 58 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFLUwWg2icaQCmqHqVt:-KFLUwWhgmBVEhhiIYHI:-595435801) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L58)):*
nit: the name confused me. Maybe `itersPerProto`?

On second thought, for optimal clarity, maybe bundle it all together?

```go
var belowRaftGoldenProtos = struct {
  fixtures map[reflect.Type]fixture
  itersPerProto int
  goldenSeed int64
} {
   ...
}
```

---

*[storage/main_test.go, line 52 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG7fezQvMyV4BbErjR:-KFLVcKcbiOzCmv4YHXv:137984126) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/main_test.go#L52)):*
I'm also not willing to push for this. It sounds good in theory, but I'm having trouble seeing where this would fit in practice.

---

*[storage/main_test.go, line 40 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFLVnTwCPu_yx5f2w5C:-KFLVnTwCPu_yx5f2w5D:97091676) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/main_test.go#L40)):*
> // We'll remove the encountered protos after the tests.

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFLWm3BTf_tPyW0Lo9x:1232472862) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
Ah, gotcha.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-:-KFLWtm56F0jMtKvWzBP:-1464290319)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Reviewed 2 of 42 files at r1, 8 of 11 files at r4, 33 of 37 files at r5, 1 of 5 files at r6, 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[kv/main_test.go, line 17 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFu85TylJw6Qlmd18x:-KFMVP-RnroJ43yc24zB:-501931562) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ecd0e64cfe8c9f47d6fde306dca59e2049007e/kv/main_test.go#L17)):*
I'm ok with just doing this testing in package storage until we uncover a need for it elsewhere (and even then, I'd prefer to expand coverage in package storage instead of adding the proto tests to other packages)

---

*[storage/below_raft_protos_test.go, line 58 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFLUwWg2icaQCmqHqVt:-KFMVjP1yWmA-dyTUdWq:1072661544) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L58)):*
Comment that we include multiple populated protos because the population function has a chance of leaving some fields blank and we want to make sure that we get examples of non-empty values for every field. 

---

*[storage/main_test.go, line 52 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG7fezQvMyV4BbErjR:-KFMUjPtUZVy5B6UVl-g:-1033579539) ([raw file](https://github.com/cockroachdb/cockroach/blob/49b2574323296ceba28d245545546e23386d62c1/storage/main_test.go#L52)):*
The fact that the different protos have different `emptySum` values satisfies me that the test is not trivially missing something; I don't think that a more artificial test is worthwhile. 

---

*[storage/main_test.go, line 73 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFMWMY0wuLRgxUZbLBG:-KFMWMY0wuLRgxUZbLBH:-154289639) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/main_test.go#L73)):*
This will be a problem when running a single test under stress; I don't think it's worthwhile. It will be very rare to remove a proto from the protected list and we can handle that manually when it comes up. 

---

*[storage/track_raft_protos.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG1c4vz42eX-AJJQHZ:-KFMSZ_Ar-wBqoK37TkA:1834221512) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L48)):*
I was suggesting this for performance. If the tests are still 30% slower than before then I think it's worth addressing now. 

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFMP4DR8sf-bUTKyuHF:1003552320) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
I don't feel strongly about the type assertion; I was just suggesting it as a way to address Peter's concerns. I'm fine with the code as-is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-:-KFMX-KeAQ7MitDUnfgl:676834233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[storage/main_test.go, line 73 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFMWMY0wuLRgxUZbLBG:-KFO0GcF9OguzX0u5sjA:207836063) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/main_test.go#L73)):*
Maybe I'm overcomplicating this, but you could have

```go
var verifyRaftProtos bool

func TestMain(m *testing.M) {
  // ...
  code := m.Run()
  if verifyRaftProtos {
     // do all the checks
  }
}

TestGoldenProto(t *testing.T) {
  verifyRaftProtos = true
}
```

That way the detection really is triggered by running a specific test (though that test will never fail directly, which is just a little odd).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Did the slowdown in the tests go away?

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[storage/below_raft_protos_test.go, line 45 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFP4V5WPzHnM54cupao:-KFP4V5WPzHnM54cupap:951941073) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L45)):*
Is this used?

---

*[storage/below_raft_protos_test.go, line 119 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFP4XKQwyeHNdLSobDl:-KFP4XKQwyeHNdLSobDm:1579449496) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L119)):*
Any reason you're using `marshalTo` here, but `protoutil.Marshal` above?  Performance isn't a concern, so it seems fine to use `protoutil.Marshal`.

---

*[storage/engine/engine.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFFn7vCROfc0jdm1PI-:-KFP5-qV2s5QGJDYc3Gg:1879950342) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c566546cbf2811b414b46ca50fb423c6a50c32/storage/engine/engine.go#L186)):*
Ditto. The code is fine as-is.

---

*[util/protoutil/marshal.go, line 23 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFP5YiUbIL4zW_46IvQ:-KFP5YiUbIL4zW_46IvR:1386148703) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/util/protoutil/marshal.go#L23)):*
Curious why you moved away from using an `atomic.Value`. Perhaps the `atomic.Value` is a bit paranoid, but was it actually causing a performance problem?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-:-KFP61Db7Q_EWQGneAlo:991800614)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[util/protoutil/marshal.go, line 23 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFP5YiUbIL4zW_46IvQ:-KFPAXxDNrBAxLy7NBQQ:954230549) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/util/protoutil/marshal.go#L23)):*
I pointed out in a (vanished?) comment that the synchronization wasn't necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

tamird commented:
Yep, see reply to @bdarnell elsewhere in this batch of comments.

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[roachpb/data.proto, line 42 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJ6aAE9ZFDjNkNYE5r:-KFNyRn-agiDIsAH6S0g:373471512) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/roachpb/data.proto#L42)):*
I don't get what you're saying; yes, I could have those methods generated for our proto types, but it wouldn't compile because they would depend on populate methods for raft's proto types.

---

*[storage/below_raft_protos_test.go, line 45 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFP4V5WPzHnM54cupao:-KFP8t8Zu5yH89tZKSAY:1100623757) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L45)):*
detritus. removed.

---

*[storage/below_raft_protos_test.go, line 58 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFLUwWg2icaQCmqHqVt:-KFPAuyiuE_6Em-2NDka:-1173281406) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L58)):*
@bdarnell done. @tschottdorf I added comments that should address your concerns - I don't want to add the same numeric constants to every entry in this map =/

---

*[storage/below_raft_protos_test.go, line 119 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFP4XKQwyeHNdLSobDl:-KFPB1V8e9eAmoqZ7ry1:784671944) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L119)):*
The reason is not so strong - I wanted to pass a single slice to `verifyHash`, and using marshalTo was more convenient than Marshal + `append`. Is it worth changing?

---

*[storage/main_test.go, line 40 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFLVnTwCPu_yx5f2w5C:-KFP7F1pC40OdqF0rcKX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/main_test.go#L40)):*
Done.

---

*[storage/main_test.go, line 73 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFMWMY0wuLRgxUZbLBG:-KFP7G0IA-BHwmOgEsRq:1061030482) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/main_test.go#L73)):*
That seems better to me. Done.

---

*[storage/track_raft_protos.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFG1c4vz42eX-AJJQHZ:-KFP7hXlIs_xEZlIBfcv:-392787021) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db8ca4f0e647d13193fcdc656112a2e6ee17ea7/storage/track_raft_protos.go#L48)):*
before: 25.743, after: 26s

I added a fast-bail since this comment was posted
```
		belowRaftProtos.Lock()
		_, ok := belowRaftProtos.inner[t]
		belowRaftProtos.Unlock()
		if ok {
			return
		}
```

which probably accounts for the lack of slowdown.

---

*[util/protoutil/marshal.go, line 23 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFP5YiUbIL4zW_46IvQ:-KFP8ZdKQHwnt1KFhx3i:-1769462939) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/util/protoutil/marshal.go#L23)):*
@tschottdorf suggested it, and I didn't have a strong reason to object, so I did it.

I don't feel strongly either way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-:-KFP8jkVE5GDHpwILx0z:-380864395)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 38 files at r10.
Review status: 1 of 44 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[roachpb/data.proto, line 42 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFJ6aAE9ZFDjNkNYE5r:-KFPDJj-QAPdV7Qr-K7U:1720531546) ([raw file](https://github.com/cockroachdb/cockroach/blob/928032e032286222c5d412f83d8a4f58e568e62d/roachpb/data.proto#L42)):*
Oh, sorry. Hadn't understood what you meant.

---

*[storage/below_raft_protos_test.go, line 58 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFLUwWg2icaQCmqHqVt:-KFPFlppc6SgndZUjSmu:-1094005622) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L58)):*
Fine by me, but note that I wasn't suggesting adding info to every fixture, but to make `belowRaftGoldenProtos` contain the magic numbers the fixtures depend on. I'm happy with the way it is, though.

---

*[util/protoutil/marshal.go, line 23 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFP5YiUbIL4zW_46IvQ:-KFPFWpMxggmmhy6d_BX:-976792377) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/util/protoutil/marshal.go#L23)):*
I'd rather not have synchronization on every proto marshal in our production code until it actually matters, but also not ready to go to war for it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 44 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/below_raft_protos_test.go, line 119 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFP4XKQwyeHNdLSobDl:-KFPGLpNsAIMb0sQ8eE0:-230949313) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L119)):*
No, probably not worth changing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 46 files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/below_raft_protos_test.go, line 58 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038#-KFLUwWg2icaQCmqHqVt:-KFPGxuu9fqohmYSNmpR:1802113536) ([raw file](https://github.com/cockroachdb/cockroach/blob/4647eefd619df2d7feb682335ba36b6b11509451/storage/below_raft_protos_test.go#L58)):*
Ah, understood (and left as is).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6038)*
<!-- Sent from Reviewable.io -->
